File: jas91_smaf91/transformation-2.js
/*
 * Transformation 2
 *
 * Description:
 *       Finds the number of crimes and 311 reports in a 100ft sqare area. Based
 *       on latitude and longitude.
 */

db.system.js.save({ _id: "transformation2", value: function() {
    db.jas91_smaf91.crimes_per_location.drop();
    db.jas91_smaf91.sr311_per_location.drop();
    db.jas91_smaf91.union_crime_sr311.drop();
    db.jas91_smaf91.sr311_crime_per_location.drop();

    // Get number of crimes per coordinates
    db.jas91_smaf91.crime.mapReduce(
        function(){
            var id = {
                latitude:  this.geo_info.geometry.coordinates[0], 
                longitude: this.geo_info.geometry.coordinates[1],
                type: 'crime'
            }
            emit(id, 1);
        },
        function(k,vs) {
            return Array.sum(vs)
        },
        { 
            out: "jas91_smaf91.crime_per_location"
        }
    );

    // Get number of 311 requests per coordinates
    db.jas91_smaf91.sr311.mapReduce(
       function(){
          var id = {
                latitude:  this.geo_info.geometry.coordinates[0], 
                longitude: this.geo_info.geometry.coordinates[1],
                type: 'sr311'
           }
            emit(id, 1);
        },
        function(k,vs) {
            return Array.sum(vs)
        },
        { 
            out: "jas91_smaf91.sr311_per_location"
        }
    );

    function union(X, Y, XY) {
        db[XY].remove({});
        db.createCollection(XY);
        db[X].find().forEach(function(x) {
           db[XY].insert(x);
        });
        db[Y].find().forEach(function(y) {
            db[XY].insert(y);
       });
    }

    // Get the union between the crimes per coordinates and 311 request per coordinates
    union('jas91_smaf91.crime_per_location', 'jas91_smaf91.sr311_per_location', 'jas91_smaf91.union_crime_sr311')

    // Join those two datasets per coordinates to get the following
    // (latitude, longitude) -> # of crimes, # of 311 requests
    db.jas91_smaf91.union_crime_sr311.mapReduce(
        function(){
            var id = {
                latitude:  this._id.latitude, 
              longitude: this._id.longitude
            }
           if (this._id.type == 'sr311') {
                emit(id, {crime: 0, sr311: this.value});
            } else {
                emit(id, {crime: this.value, sr311: 0});
            }
        },
        function(k,vs) {
            var total_crimes = 0
            var total_sr311 = 0
            vs.forEach(function(v, i) {
                total_crimes += v.crime;
                total_sr311 += v.sr311;
            });
            return {crime: total_crimes, sr311: total_sr311}
        },
        { 
            out: "jas91_smaf91.sr311_crime_per_location"
        }
    );

    db.jas91_smaf91.crimes_per_location.drop();
    db.jas91_smaf91.sr311_per_location.drop();
    db.jas91_smaf91.union_crime_sr311.drop();
}});

File: jas91_smaf91/transformation2.py
import json
import dml
import prov.model
import datetime
import uuid

class transformation2(dml.Algorithm):
    contributor = 'jas91_smaf91'
    reads = ['jas91_smaf91.311', 'jas91_smaf91.hospitals', 'jas91_smaf91.food', 'jas91_smaf91.schools']
    writes = ['jas91_smaf91.crime']

    @staticmethod
    def execute(trial = False):
        startTime = datetime.datetime.now()

        # Set up the database connection.
        client = dml.pymongo.MongoClient()
        repo = client.repo
        repo.authenticate('jas91_smaf91', 'jas91_smaf91')

        def project(document):
            return {'geo_info': document['geo_info']}

        def union(collection1, collection2, result, f):
            for document in repo[collection1].find():
                document = f(document) 
                coordinates = document['geo_info']['geometry']['coordinates']
                if coordinates[0] and coordinates[1]:
                    repo[result].insert(f(document))

            for document in repo[collection2].find():
                document = f(document)
                coordinates = document['geo_info']['geometry']['coordinates']
                if coordinates[0] and coordinates[1]:
                    repo[result].insert(f(document))

        repo.dropPermanent('jas91_smaf91.union_temp')
        repo.createPermanent('jas91_smaf91.union_temp')
        print('[OUT] computing the union of the datasets')
        union('jas91_smaf91.sr311', 'jas91_smaf91.food', 'jas91_smaf91.union_temp', project)
        union('jas91_smaf91.schools', 'jas91_smaf91.hospitals', 'jas91_smaf91.union_temp', project)
        print('[OUT] done')

        print('[OUT] indexing by latitude and longitude')
        repo.jas91_smaf91.union_temp.ensure_index([('geo_info.geometry', dml.pymongo.GEOSPHERE)])
        print('[OUT] done')

        print('[OUT] filling crime zip codes')
        for document in repo.jas91_smaf91.crime.find():
            latitude = document['geo_info']['geometry']['coordinates'][0]
            longitude = document['geo_info']['geometry']['coordinates'][1]
            neighbor = repo.jas91_smaf91.union_temp.find_one({ 
                'geo_info.geometry': {
                    '$near': { 
                        '$geometry': {
                            'type': 'Point', 
                            'coordinates' :[latitude,longitude]
                            }, 
                        '$maxDistance': 1000, 
                        '$minDistance': 0 
                        } 
                    }
                })

            if neighbor:
                zip_code = neighbor['geo_info']['properties']['zip_code']
                document['geo_info']['properties']['zip_code'] = zip_code
                repo.jas91_smaf91.crime.update(
                        {'_id': document['_id']}, 
                        document,
                        upsert=False
                        )
        repo.logout()
        print('[OUT] done')

        endTime = datetime.datetime.now()

        return {"start":startTime, "end":endTime}

    @staticmethod
    def provenance(doc = prov.model.ProvDocument(), startTime = None, endTime = None):
        '''
        Create the provenance document describing everything happening
        in this script. Each run of the script will generate a new
        document describing that invocation event.
        '''

        client = dml.pymongo.MongoClient()
        repo = client.repo
        repo.authenticate('jas91_smaf91', 'jas91_smaf91')

        doc.add_namespace('alg', 'http://datamechanics.io/algorithm/') # The scripts are in <folder>#<filename> format.
        doc.add_namespace('dat', 'http://datamechanics.io/data/') # The data sets are in <user>#<collection> format.
        doc.add_namespace('ont', 'http://datamechanics.io/ontology#') # 'Extension', 'DataResource', 'DataSet', 'Retrieval', 'Query', or 'Computation'.
        doc.add_namespace('log', 'http://datamechanics.io/log/') # The event log.
        doc.add_namespace('bdp', 'https://data.cityofboston.gov/resource/')

        this_script = doc.agent('alg:jas91_smaf91#transformation2', {prov.model.PROV_TYPE:prov.model.PROV['SoftwareAgent'], 'ont:Extension':'py'})
        populate = doc.activity('log:uuid'+str(uuid.uuid4()), startTime, endTime, {'prov:label':'Populate crime zip code field'})
        doc.wasAssociatedWith(populate, this_script)

        resource_sr311 = doc.entity('dat:jas91_smaf91#sr311', {'prov:label':'311 Service Reports', prov.model.PROV_TYPE:'ont:DataSet'})
        doc.usage(
            populate, 
            resource_sr311, 
            startTime, 
            None,
            {
                prov.model.PROV_TYPE:'ont:Query',
                'ont:Query':'db.jas91_smaf91.sr311.find({},{geo_info: 1})'
            }
        )

        resource_hospitals = doc.entity('dat:jas91_smaf91#hospitals', {'prov:label':'Hospital Locations', prov.model.PROV_TYPE:'ont:DataSet'})
        doc.usage(
            populate, 
            resource_hospitals, 
            startTime, 
            None,
            {
                prov.model.PROV_TYPE:'ont:Query',
                'ont:Query':'db.jas91_smaf91.hospitals.find({},{geo_info: 1})'
            }
        )

        resource_food = doc.entity('dat:jas91_smaf91#food', {'prov:label':'Food Establishment Inspections', prov.model.PROV_TYPE:'ont:DataSet'})
        doc.usage(
            populate, 
            resource_food, 
            startTime, 
            None,
            {
                prov.model.PROV_TYPE:'ont:Query',
                'ont:Query':'db.jas91_smaf91.food.find({},{geo_info: 1})'
            }
        )

        resource_schools = doc.entity('dat:jas91_smaf91#schools', {'prov:label':'Schools', prov.model.PROV_TYPE:'ont:DataSet'})
        doc.usage(
            populate, 
            resource_schools, 
            startTime, 
            None,
            {
                prov.model.PROV_TYPE:'ont:Query',
                'ont:Query':'db.jas91_smaf91.sr311.schools({},{geo_info: 1})'
            }
        )

        crime = doc.entity('dat:jas91_smaf91#crime', {'prov:label':'Crime Incident Reports', prov.model.PROV_TYPE:'ont:DataSet'})
        doc.usage(
            populate, 
            crime, 
            startTime, 
            None,
            {prov.model.PROV_TYPE:'ont:Computation'}
        )

        doc.wasAttributedTo(crime, this_script)
        doc.wasGeneratedBy(crime, populate, endTime)

        repo.record(doc.serialize()) # Record the provenance document.
        repo.logout()

        return doc

'''
transformation2.execute()
doc = transformation2.provenance()
print(json.dumps(json.loads(doc.serialize()), indent=4))
'''
